gradle-repository/buildSrc/src/main/kotlin/gradlebuild.gradle-api-repository.gradle.kts

```diff
@@ -0,0 +1,126 @@
+plugins {
+    base
+    `maven-publish`
+}
+
+val gradleApiRepository = extensions.create<GradleRepositoryExtension>("gradleApiRepository")
+
+gradleApiRepository.apply {
+    gradleVersion.convention(
+        providers.gradleProperty("gradleApiVersion").orElse("7.4.1")
+    )
+    repoLocation.convention(
+        layout.buildDirectory.dir("repo")
+    )
+}
+
+repositories {
+    listOf("distributions", "distributions-snapshots").forEach { distUrl ->
+        ivy {
+            name = "Gradle distributions"
+            url = uri("https://services.gradle.org")
+            patternLayout {
+                artifact("/${distUrl}/[module]-[revision]-bin(.[ext])")
+            }
+            metadataSources {
+                artifact()
+            }
+            content {
+                includeModule("gradle-dist", "gradle")
+            }
+        }
+    }
+}
+
+val artifactType = Attribute.of("artifactType", String::class.java)
+val runtimeAttribute = objects.named<Usage>(Usage.JAVA_RUNTIME)
+
+val gradleDistribution by configurations.creating
+val gradleClasspath by configurations.creating {
+    extendsFrom(gradleDistribution)
+    attributes.attribute(artifactType, "gradle-classpath")
+}
+val gradleJarsConfiguration by configurations.creating {
+    extendsFrom(gradleDistribution)
+    attributes.attribute(artifactType, "gradle-libs-dir")
+}
+val gradleArtifactListConfiguration by configurations.creating {
+    extendsFrom(gradleDistribution)
+    attributes.attribute(artifactType, "gradle-artifact-list")
+}
+
+dependencies {
+    gradleDistribution(gradleApiRepository.gradleVersion.map { "gradle-dist:gradle:${it}@zip" })
+
+    registerTransform(ExplodeZipAndFindJars::class) {
+        from.attribute(artifactType, "zip")
+        to.attribute(artifactType, "gradle-libs-dir")
+    }
+    registerTransform(CreateGradleArtifactList::class) {
+        parameters {
+            gradleVersion.set(gradleApiRepository.gradleVersion)
+        }
+        from.attribute(artifactType, "gradle-libs-dir")
+        to.attribute(artifactType, "gradle-artifact-list")
+    }
+}
+
+val thirdPartyJarsPath = "third-party"
+val gradleJarsPath = "gradle-jars"
+
+val resolveJars by tasks.registering(Sync::class) {
+    from(gradleJarsConfiguration) {
+        into(thirdPartyJarsPath)
+        exclude("gradle-*")
+    }
+    from(gradleJarsConfiguration) {
+        into(gradleJarsPath)
+        include("gradle-*")
+    }
+    into(project.buildDir.resolve("libs"))
+}
+
+afterEvaluate {
```

Comment:
Why is `afterEvaluate` needed here?

Reply:
So I see publishing needs that. What part of `publishing` doesn't work here?

Reply:
I left a comment explaining why we need the `afterEvaluate`.